compute-hypothetical-feature.mjs

```diff
@@ -0,0 +1,143 @@
+    return snapshots.at(index);
+}
+
+const months = Array.from(iterateMonths(Date.parse('2020-01-01T00:00Z'), endDate));
+
+function getVersionAfter(browser, date) {
+    for (const release of caniuse.agents[browser].version_list) {
+        if (!release.release_date) {
+            continue;
+        }
+        const releaseDate = caniuseDate(release.release_date);
+        if (releaseDate >= date) {
+            return release.version;
+        }
+    }
+    return null;
+}
+
+for (const startIndex of months.keys()) {
+    const launchDate = months[startIndex];
+    // Assume that a hypothetical feature is enabled in all browser engines at
+    // |launchDate| and is available in all following browser releases. Step
+    // forward one month at a time and compute availability based on the
+    // historical usage stats in |historicalBrowserData|.
+
+    // Identify the version of each browser where the feature was first available.
+    // This is done up front to make it simpler to determine if each release has
+    // the feature based on the version, not the release date.
+    const launchVersions = {};
+    for (const browser of browsers) {
+        launchVersions[browser] = getVersionAfter(browser, launchDate);
+    }
+
+    console.log(monthString(launchDate));
+    for (let index = startIndex; index < months.length; index++) {
+        const month = months[index];
+
+        let totalUsage = 0;
+        let totalAvailability = 0;
+
+        let missingData = false;
+        for (const browser of browsers) {
+            const launchVersion = launchVersions[browser];
+            if (!launchVersion) {
+                missingData = true;
+                break;
+            }
+
+            const historicalUsage = getHistoricalUsage(browser, month);
+            if (!historicalUsage) {
+                missingData = true;
+                break;
+            }
+
+            // TODO check freshness of |historicalUsage|.
+            //console.debug(`  Usage data from ${caniuseDate(historicalUsage.timestamp).toISOString()}`);
+
+            let browserUsage = 0;
+            let browserAvailability = 0;
+            for (const [version, usage] of Object.entries(historicalUsage.usage_global)) {
+                browserUsage += usage;
+                let support = false;
+                if (caniuseCompare(version, launchVersion, '>=')) {
+                    support = true;
+                    browserAvailability += usage;
+                }
+                //console.debug(browser, version, usage, support ? '(included)' : '(not included)');
+            }
+            totalUsage += browserUsage;
+            totalAvailability += browserAvailability;
+            console.log(`  -> ${monthString(month)}: ${browser} availability ${(100 * browserAvailability / browserUsage).toFixed(2)}% (${browserAvailability.toFixed(2)}% / ${browserUsage.toFixed(2)}%)`);
+        }
+        if (!missingData) {
```

Comment:
Yes this is ugly. Could break out the body of `let index` loop into a function and just do early returns instead.